netket/optimizer/sr/sr_jacobian.py

```diff
@@ -0,0 +1,142 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Optional, Union, Tuple, Any
+from functools import partial
+
+import jax
+import flax
+from jax import numpy as jnp
+from flax import struct
+
+from netket.utils.types import PyTree, Array
+from netket.utils import rename_class
+import netket.jax as nkjax
+
+from .s_jacobian_mat import JacobianSMatrix, gradients
+
+from .base import SR
+
+
+@struct.dataclass
+class SRJacobian(SR):
+    """
+    Base class holding the parameters for the iterative solution of the
+    SR system x = ⟨S⟩⁻¹⟨F⟩, where S is a lazy linear operator
+
+    Tolerances are applied according to the formula
+    :code:`norm(residual) <= max(tol*norm(b), atol)`
+    """
+
+    tol: float = 1.0e-5
+    """Relative tolerance for convergences."""
+
+    atol: float = 0.0
+    """Absolutes tolerance for convergences."""
+
+    maxiter: int = None
+    """Maximum number of iterations. Iteration will stop after maxiter steps even 
+    if the specified tolerance has not been achieved.
+    """
+
+    M: Optional[Union[Callable, Array]] = None
+    """Preconditioner for A. The preconditioner should approximate the inverse of A. 
+    Effective preconditioning dramatically improves the rate of convergence, which implies 
+    that fewer iterations are needed to reach a given error tolerance.
+    """
+
+    mode: str = struct.field(pytree_node=False)
+    """Differentiation mode to precompute Jacobian
+    * "holomorphic": C->C holomorphic function
+        `grad` is called on the full network output with `holomorphic=True`
+    * "R2R": real-valued wave function with real parameters
+        `grad` is called on the real part of the network output with `holomorphic=False`
+    * "R2C": complex-valued wave function with real parameters
+        the real and imaginary parts of the network output are treated as independent 
+        R->R functions and `grad` is called separately on them with `holomorphic=False`
+    """
+
+    rescale_shift: bool = struct.field(pytree_node=False, default=False)
+    """Whether scale-invariant regularisation should be used"""
+
+    def create(self, *args, **kwargs):
+        if rescale_shift:
+            O, scale = gradients(apply_fun, params, samples, model_state, self.mode, True)
+            return JacobianSMatrix(sr=sr, x0=x0, O=O, scale=scale)
+        else:
+            O = gradients(apply_fun, params, samples, model_state, self.mode, False)
+            return JacobianSMatrix(sr=sr, x0=x0, O=O)
+
```

Comment:
And with the change above this could became...
(I think it's a bit clearer)
```suggestion
        O, scale = gradients(apply_fun, params, samples, model_state, self.mode, rescale_shift)
        return JacobianSMatrix(sr=sr, x0=x0, O=O, scale=scale)
```

Reply:
nice, this is much cleaner!